### ISSUE TITLE
How can I create customized Numpylevel layer(operator) such as FullyConnect

### ISSUE THREAD

**Issue by hehaodele** [dmlc/mxnet#879]
I want to build a sparse connected layer in MXNET by myself, but there is no example for how to build a layer having weight and need to calculate gradient? Can anyone help? Thanks very much^_^

**Comment by jermainewang** [comment]
Hi, could you try [numpy-level op](https://github.com/dmlc/mxnet/tree/master/example/numpy-ops) to do that?

**Comment by hehaodele** [comment]
@jermainewang Following is the code I used to build a sparse connected layer
```
class SingleConnected(mx.operator.NumpyOp):
    def __init__(self):
        super(SingleConnected, self).__init__(True)

    def list_arguments(self):
        return ['data', 'weight', 'bias']

    def list_outputs(self):
        return ['output']

    def infer_shape(self, in_shape):
        data_shape = in_shape[0]
        weight_shape = (in_shape[0][1],)
        bias_shape = (in_shape[0][1],)
        output_shape = in_shape[0]
        return [data_shape, weight_shape, bias_shape], [output_shape]

    def forward(self, in_data, out_data):
        x = in_data[0]
        w = in_data[1]
        b = in_data[2]
        y = out_data[0]
        # element-wise
        y = x * w + b

    def backward(self, out_grad, in_data, out_data, in_grad):
        dx = in_grad[0]
        dw = in_grad[1]
        db = in_grad[2]
        dy = out_grad[0]
        x  = in_data[0]
        w  = in_data[1]
        dx[:] = dy * w
        dw[:] = np.sum(dy * x, axis=0)
        db[:] = np.sum(dy, axis=0)
        print 'dy', dy
```
In addition, the structure and model are defined as following
```
data = mx.symbol.Variable('data')
mySingleConnected = SingleConnected()
hh = mySingleConnected(data = data, name='hh')
fc = mx.symbol.FullyConnected(data = hh, name='fc', num_hidden = 1)
mlp = mx.symbol.LinearRegressionOutput(data = fc, name = 'regression')


model = mx.model.FeedForward(
    ctx = mx.cpu(), symbol = mlp, num_epoch = 2,
    learning_rate = 0.001, momentum = 0.9, wd = 0.,
    # optimizer          = 'adam',
    initializer        = mx.init.Uniform(0.001)
)
```
I print ```dy``` during training, yet the printed value sometimes become ```inf``` or ```nan```. Any ideas to solve this? Thanks!

**Comment by piiswrong** [comment]
You said "sometimes become inf or nan" so it's not always inf or nan? This most likely is because your learning rate is too big

**Comment by hehaodele** [comment]
I have tried very small learning rate and small parameter initial factor, but the problem still have. SO I wonder whether it is because, I do not build the layer in the correct way. For example, I do not understand the accurate definition of the```in_shape```, why it is a list, I find that the ```in_shape[0]``` is a tuple of ```(batch_size, dimension_of_data)```, but what is the meaning of ```in_shape[1]```
following is the data I used to train the model:
```
data            = np.zeros((100,2))
# data[0:5,:]    = np.zeros((5,3)) - 1
# data[5:10,:]  = np.zeros((5,3)) + 1
data[0:50,:]    = np.random.randn(50,3) - 2
data[50:100,:]  = np.random.randn(50,3) + 2

label           = np.zeros((100,1))
label[0:50]     = -1
label[50:100]   = 1


train = mx.io.NDArrayIter(
    data = data,
    label = label.flatten(),
    batch_size = 100,
    shuffle=False
)
```

**Comment by piiswrong** [comment]
Please refer to https://mxnet.readthedocs.org/en/latest/tutorial/new_op_howto.html

**Comment by hehaodele** [comment]
Well, I think my layer is consistent to the reference. Here is another question: If I want to initialize my layer with 'weight' equals one and 'bias' equal zero, but the other layer(provided by mxnet) with mx.init.Uniform(0.001). Is there any easy way to implement this? or it is not supported by MXNET. Thanks very much~~

**Comment by cuihengbin** [comment]
@jermainewang @piiswrong
Additionally, if I want to set the "num_hidden", how should I code in my new created layer? 
Maybe someone can show a complete example of fully-connected layer with numpylevel/ndarraylevel layer operator.
Needs your help ~ Thank you very much.

**Comment by tqchen** [comment]
@hehaodele  you can create initializer by name of the weights, see also https://github.com/dmlc/mxnet/issues/902
